sql: add crdb_internal.protect_cluster builtin

stevendanna opened the pull request:
Migration tooling uses historical queries that may need to run for many minutes or hours. Ensuring that these queries can continue to completion requires protecting the table's data from garbage collection.

The new builtin crdb_internal.protect_cluster creates a cluster-wide PTS record at the given timestamp. The timestamp will expire in 24 hours or when the returned job ID is canceled.

We re-purpose the stream ingestion producer job as the owner of this timestamp. This has a couple of advantages over a session-scoped builtin for now:

- Ability for an operator to remove the PTS record explicitly using CANCEL JOB.

- Some visibility into the source of the PTS record via the jobs table.

We use a cluster-wide PTS record rather than a table-specific PTS record because to be sure that we can do a historical query as of a given timestamp also requires that we have access to at least the descriptor and namespace table at that timestamp.

We've re-used the stream ingestion producer job for now as it also gives a simple way for the caller to extend the PTS record beyond 24 hours with crdb_internal.replication_stream_progress(job_id, protectTS).

Epic: CC-27068

Release note: None

stevendanna committed:
sql: add crdb_internal.protect_cluster builtin

Migration tooling uses historical queries that may need to run for
many minutes or hours. Ensuring that these queries can continue to
completion requires protecting the table's data from garbage
collection.

The new builtin crdb_internal.protect_cluster creates a cluster-wide
PTS record at the given timestamp. The timestamp will expire in 24
hours or when the returned job ID is canceled.

We re-purpose the stream ingestion producer job as the owner of this
timestamp. This has a couple of advantages over a session-scoped
builtin for now:

- Ability for an operator to remove the PTS record explicitly using
  CANCEL JOB.

- Some visibility into the source of the PTS record via the jobs
  table.

We use a cluster-wide PTS record rather than a table-specific PTS
record because to be sure that we can do a historical query as of a
given timestamp also requires that we have access to at least the
descriptor and namespace table at that timestamp.

We've re-used the stream ingestion producer job for now as it also
gives a simple way for the caller to extend the PTS record beyond 24
hours with crdb_internal.replication_stream_progress(job_id,
protectTS).

Epic: CC-27068

Release note: None

stevendanna commented on pkg/ccl/streamingccl/streamproducer/stream_lifetime.go:
```diff
@@ -147,6 +148,39 @@ func StartReplicationProducerJob(
 
+// StartHistoryProtectionJob replication stream producer job
+// responsible for managing a cluster-level protected timestamp.
+func StartHistoryProtectionJob(
```
We can avoid alot of the indirection here if we are OK moving the producer job resumer to `pkg/repstream`. That would make it available in non-CCL binaries. Then we would move StartHistoryProtectionJob to just live on sql.Planner. If this seems cool with y'all let me know.

I don't think the job's resumer particularly needs to be restricted to CCL given that we would still be restricting many of the builtins to CCL code.

ajstorm reviewed (approved):
This looks great to me, thanks @stevendanna!  Left a few comments, but you'll probably also want a jobs expert to take a look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120727#-:-NtRmWXeDqBmbDCq59kk:b-s47qho)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamproducer/stream_lifetime.go` line 161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120727#-NtRkg1jEwGPdLJlm38d:-NtRkg1jEwGPdLJlm38e:b6lm1pt) ([raw file](https://github.com/cockroachdb/cockroach/blob/5816b0b5d4375c50d7b768b93c227b7d7a6dc8c3/pkg/ccl/streamingccl/streamproducer/stream_lifetime.go#L161)):*
> ```Go
> 
> 	jr := makeJobRecordForClusterPTSRetention(registry,
> 		defaultExpirationWindow,
> ```

Nit: Does it make sense to allow this to be modifiable through a cluster setting?  My vague concern is that we encounter problems with it either being too large or too small in the field and can't get out of it without some pain.  It's a very remote concern though, so not at all essential to address on this PR
___
*[`pkg/ccl/logictestccl/testdata/logic_test/builtins` line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120727#-NtRlr0eCaZo3R6XxtBD:-NtRlr0eCaZo3R6XxtBE:b76h33h) ([raw file](https://github.com/cockroachdb/cockroach/blob/5816b0b5d4375c50d7b768b93c227b7d7a6dc8c3/pkg/ccl/logictestccl/testdata/logic_test/builtins#L28)):*
> ```
> 
> query T
> SELECT description FROM [SHOW JOBS] WHERE job_type = 'REPLICATION STREAM PRODUCER'
> ```

Nit: how hard would it be for us to have a new HISTORY PROTECTION job type?  Not a deal-breaker, especially this late in the cycle, but this job type is a bit misleading.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120727)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @dt)

___
*[`pkg/ccl/streamingccl/streamproducer/stream_lifetime.go` line 161 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120727#-NtRkg1jEwGPdLJlm38d:-NtS4B9m4kwUQJ2OZrUl:baof3ga) ([raw file](https://github.com/cockroachdb/cockroach/blob/5816b0b5d4375c50d7b768b93c227b7d7a6dc8c3/pkg/ccl/streamingccl/streamproducer/stream_lifetime.go#L161)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: Does it make sense to allow this to be modifiable through a cluster setting?  My vague concern is that we encounter problems with it either being too large or too small in the field and can't get out of it without some pain.  It's a very remote concern though, so not at all essential to address on this PR
</blockquote></details>

I think I would mildly prefer that we just allow the user to pass it in rather than have a cluster setting.  So MOLT could pass in whatever value is appropriate and it could be configurable there (perhaps with some very high upper bound to what the user passes in).

If we have a PTS with a long expiration that we need to free before that expiration hits: `CANCEL JOB $JOB_ID`
If we have a PTS with a short expiration that we need to extend: `select crdb_internal.replication_stream_progress($JOB_ID, $PROTECT_TS)`
___
*[`pkg/ccl/logictestccl/testdata/logic_test/builtins` line 28 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/120727#-NtRlr0eCaZo3R6XxtBD:-NtS5-V15mKib6Yl-r9f:b3gcmo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5816b0b5d4375c50d7b768b93c227b7d7a6dc8c3/pkg/ccl/logictestccl/testdata/logic_test/builtins#L28)):*
<details><summary><i>Previously, ajstorm (Adam Storm) wrote…</i></summary><blockquote>

Nit: how hard would it be for us to have a new HISTORY PROTECTION job type?  Not a deal-breaker, especially this late in the cycle, but this job type is a bit misleading.
</blockquote></details>

Not too hard. One small nice thing about reusing this job type is that we get to reuse the progress update functions for heartbeating. But perhaps that's worth the tradeoff.


<!-- Sent from Reviewable.io -->

stevendanna committed:
sql,streamingccl: make history protection expiration configurable

Release note: None
Epic: none

stevendanna committed:
sql: add HISTORY RETENTION job

Previously, crdb_internal.protect_cluster re-used the existing STREAM
PRODUCER JOB type. Here we make a new job type so that the `SHOW JOBS`
output is a bit cleaner. It also means that this function no longer
depends on any CCL code.

Epic: none
Release note: None

ajstorm commented:
These changes LGTM.  @dt were you planning on having a look too?

dt committed:
builtins: rename protect_cluster to protect_mvcc_history

Release note: none.
Epic: none.

dt commented:
bors r=dt

Jeremyyang920 commented:
How far are we able to backport this to? Just interested to know incase we need to put into our fetch code a version gate based on CRDB source version or if it can be backported to all supported versions.

dt commented:
we cannot backport a new job type

Jeremyyang920 commented:
Got it. Thanks.

ajstorm commented:
> we cannot backport a new job type

Is it possible to backport the first commit only, and deal with the existing job type in older releases?

stevendanna commented:
Uff.  We should fix this going forward because it looks to me like this would be safe if it weren't for this one function:

```golang
// Type returns the payload's job type and panics if the type is invalid.
func (p *Payload) Type() Type {
	typ, err := DetailsType(p.Details)
	if err != nil {
		panic(err)
	}
	return typ
}
```

The other places where we panic are called by nodes _writing_ the new job or updating the job record, which by definition know how to process the job type.  Perhaps there are more that I missed but Type() looks like one of the larger issues to me.  

But that also implies that this PR is incorrect without a version check before adding the job.

> Is it possible to backport the first commit only, and deal with the existing job type in older releases?

Probably possible to do the first two commits and perhaps the rename. Might not be the cleanest backport.
